File: src/shared/overlay/stage-size.ts
/**
 * How many pixels an editor stage actually works on.
 *
 * The stage used to draw at the media's own density, which is right for a clip
 * — 4K is 8.3 megapixels — and ruinous for a photograph: a 48-megapixel JPEG
 * off a drone put a 194 MB canvas on screen, and the moment a LUT was picked
 * the grader added a texture and a drawing buffer of the same size on the GPU.
 * On an iPhone that is the whole budget and the tab is killed.
 *
 * So the stage has a pixel budget: no more than a 4K frame, which leaves every
 * clip the studio was built for untouched and only bites on a big still. It is
 * a PREVIEW budget and nothing else — the exports and the frame grab compose
 * from the source bitmap at its own density, never from this canvas.
 */

/** A 4K frame's worth of pixels — the most any stage will work on. */
export const MAX_STAGE_PIXELS = 3840 * 2160;

/**
 * The frame size a stage should hold `w`×`h` in: the media's own size while it
 * is within the budget, scaled down by area — aspect kept — once it is not.
 */
export function stageFrameSize(
  w: number,
  h: number,
  budget = MAX_STAGE_PIXELS,
): { w: number; h: number } {
  if (!(w > 0) || !(h > 0)) return { w: 0, h: 0 };
  const pixels = w * h;
  if (!(budget > 0) || pixels <= budget) {
    return { w: Math.round(w), h: Math.round(h) };
  }
  const scale = Math.sqrt(budget / pixels);
  return {
    w: Math.max(1, Math.round(w * scale)),
    h: Math.max(1, Math.round(h * scale)),
  };
}

File: apple/Packages/AtelierKit/Sources/AtelierKit/Overlay/StageSize.swift
// How many pixels an editor stage works on — port of
// `src/shared/overlay/stage-size.ts`, pure.
//
// The stage used to draw at the media's own density, right for a clip (4K is
// 8.3 megapixels) and ruinous for a photograph: a 48-megapixel still put a
// 194 MB buffer on screen and killed the tab on an iPhone. So the stage has a
// PIXEL BUDGET, no more than a 4K frame — every clip the Studio was built for
// is untouched, only a big still is scaled. A PREVIEW budget and nothing else:
// exports and the frame grab compose from the source at its own density.

import Foundation

/// A 4K frame's worth of pixels — the most any stage will work on.
public let maxStagePixels = 3840.0 * 2160.0

/// JS `Math.round`: halves toward +∞.
private func jsRound(_ x: Double) -> Double {
    let floor = x.rounded(.down)
    return x - floor >= 0.5 ? floor + 1 : floor
}

/// The frame size a stage should hold `w`×`h` in: the media's own size while
/// it is within the budget, scaled down by AREA — aspect kept — once it is
/// not. 0×0 for a frame that has no size yet.
public func stageFrameSize(_ w: Double, _ h: Double, budget: Double = maxStagePixels) -> Size {
    guard w > 0, h > 0 else { return Size(0, 0) }
    let pixels = w * h
    if !(budget > 0) || pixels <= budget { return Size(jsRound(w), jsRound(h)) }
    let scale = (budget / pixels).squareRoot()
    return Size(max(1, jsRound(w * scale)), max(1, jsRound(h * scale)))
}
